src/hssm/plotting/utils.py

```diff
@@ -200,3 +239,154 @@ def _get_title(cols: Iterable["str"], col_values: Iterable[Any]) -> str:
+    row: str | None,
+    col: str | None,
+) -> tuple[Iterable[str], dict[str, Iterable[str]]]:
+    """Check the validity of `groups` and `groups_order`."""
+    if groups is None:
+        if groups_order is not None:
+            raise ValueError("`group_order` is only valid when `group` is provided.")
+    else:
+        if row is None or col is None:
+            raise ValueError(
+                "When `group` is provided, both `row` and `col` must be provided."
+            )
+        if groups_order is not None:
+            if isinstance(groups_order, Iterable) and not isinstance(
+                groups_order, dict
+            ):
+                if not isinstance(groups, str):
+                    raise ValueError(
+                        "`groups_order` can be a List-like only when `groups` is a str."
+                    )
+                groups_order = {groups: groups_order}
+            elif isinstance(groups_order, dict):
+                if not set(groups_order.keys()).issubset(set(groups)):
+                    raise ValueError(
+                        "`groups_order` can only contain keys that are in `groups`."
+                    )
+        else:
+            groups_order = {}
+        if isinstance(groups, str):
+            groups = [groups]
+
+    # Cast to the right types to satisfy mypy
+    groups = cast(Iterable, groups)
+    groups_order = cast(dict, groups_order)
+
+    return groups, groups_order
+
+
+def _use_traces_or_sample(
+    model,
+    data: pd.DataFrame | None,
+    idata: az.InferenceData | None,
+    n_samples: int | float | None,
+) -> tuple[az.InferenceData, bool]:
+    """Check if idata is provided, otherwise use traces.
+
+    Also, if posterior predictive samples are not contained in traces, sample from
+    the model.
+    """
+    # First, determine whether posterior predictive samples are available
+    # If not, we need to sample from the posterior
+    if idata is None:
+        if model.traces is None:
+            raise ValueError(
+                "No InferenceData object provided. Please provide an InferenceData "
+                + "object or sample the model first using model.sample()."
+            )
+        idata = model.traces
+
+    sampled = False
+
+    if "posterior_predictive" not in idata:
+        _logger.info(
+            "No posterior predictive samples found. Generating posterior predictive "
+            + "samples using the provided InferenceData object and the original data. "
+            + "This will modify the provided InferenceData object, or if not provided, "
+            + "the traces object stored inside the model."
+        )
+        model.sample_posterior_predictive(
+            idata=idata,
+            data=data,
+            inplace=True,
+            n_samples=n_samples,
+        )
+        idata = model.traces
+        sampled = True
+
+    return cast(az.InferenceData, idata), sampled
```

Comment:
is this not coming back as inference data?
could find it immediately, but the `sampled` return here is just for testing or is that used somewhere in the calling function?

Reply:
It is coming out as `InferenceData`, but since `idata` is defined as `InferenceData | None` in function signature, `mypy` thinks the return type here is also `InferenceData | None` which isn't consistent the type annotation, so we need to explicitly tell it that it cannot be `None`

And yes, `sampled` here just indicates whether to skip certain steps downstream